pkg/controllers/federatedcluster/plugins/katalyst.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2023 The KubeAdmiral Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugins
+
+import (
+	"context"
+
+	katalystv1a1 "github.com/kubewharf/katalyst-api/pkg/apis/node/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/kubewharf/kubeadmiral/pkg/util/logging"
+	"github.com/kubewharf/kubeadmiral/pkg/util/resource"
+)
+
+var katalystCNR = katalystv1a1.SchemeGroupVersion.WithResource(katalystv1a1.ResourceNameKatalystCNR)
+
+type katalystPlugin struct{}
+
+func (k *katalystPlugin) CollectClusterResources(
+	ctx context.Context,
+	nodes []*corev1.Node,
+	pods []*corev1.Pod,
+	handle ClusterHandle,
+) (allocatable, available corev1.ResourceList, err error) {
+	_, logger := logging.InjectLoggerValues(ctx, "gvr", katalystCNR)
+
+	cnr, ok := handle.DynamicLister[katalystCNR]
+	if !ok {
+		logger.V(4).Info("Lister not found, cluster resource collection skipped")
+		return nil, nil, nil
+	}
+	unsList, err := cnr.List(labels.Everything())
+	if err != nil || len(unsList) == 0 {
+		return nil, nil, err
+	}
+
+	allocatable = make(corev1.ResourceList)
+	nodeSet := sets.New[string]()
+	for _, node := range nodes {
+		nodeSet.Insert(node.Name)
+	}
+
+	for _, uns := range unsList {
+		unsCNR, ok := uns.(*unstructured.Unstructured)
+		if !ok || !nodeSet.Has(unsCNR.GetName()) {
```

Comment:
1. Yes, CNR needs to match the nodes of the cluster, and there will be a agent runs in every katalyst enabled node which maintains it's CNR resource
2. Currently, CNR do not support taints, but the katalyst scheduler will identify the taints on native nodes, and we already have filters for these. I think the virtual node does not support the installation of katalyst's agent because its resources are too large and there is no need to distinguish between online and offline resources.